File: src/main/java/me/senseiwells/arucas/utils/impl/ArucasOrderedMap.java
package me.senseiwells.arucas.utils.impl;

import me.senseiwells.arucas.throwables.CodeError;
import me.senseiwells.arucas.utils.Context;
import me.senseiwells.arucas.utils.ValuePair;
import me.senseiwells.arucas.values.Value;

import java.util.LinkedHashSet;
import java.util.Set;

// LinkedHashMap
public class ArucasOrderedMap extends ArucasMap {
	private Entry head;
	private Entry tail;
	private final boolean accessOrder;

	public ArucasOrderedMap() {
		super();
		this.accessOrder = false;
	}

	public ArucasOrderedMap(Context context, ArucasMap arucasMap) throws CodeError {
		this();
		this.putMapEntries(context, arucasMap, false);
	}

	private void linkNodeLast(Entry entry) {
		synchronized (this.LOCK) {
			Entry last = this.tail;
			this.tail = entry;
			if (last == null) {
				this.head = entry;
				return;
			}
			entry.before = last;
			last.after = entry;
		}
	}

	private void transferLinks(Entry source, Entry destination) {
		synchronized (this.LOCK) {
			Entry b = destination.before = source.before;
			Entry a = destination.after = source.after;
			if (b == null) {
				this.head = destination;
			}
			else {
				b.after = destination;
			}
			if (a == null) {
				this.tail = destination;
			}
			else {
				a.before = destination;
			}
		}
	}

	@Override
	Node newNode(int hash, Value<?> key, Value<?> value, TreeNode next) {
		Entry entry = new Entry(hash, key, value, next);
		this.linkNodeLast(entry);
		return entry;
	}

	@Override
	Node replacementNode(Node node, Node next) {
		Entry q = (Entry) node;
		Entry t = new Entry(q.hash, q.key, q.value, next);
		this.transferLinks(q, t);
		return t;
	}

	@Override
	TreeNode newTreeNode(int hash, Value<?> key, Value<?> value, Node next) {
		TreeNode p = new TreeNode(hash, key, value, next);
		this.linkNodeLast(p);
		return p;
	}

	@Override
	TreeNode replacementTreeNode(Node node, Node next) {
		Entry q = (Entry) node;
		TreeNode t = new TreeNode(q.hash, q.key, q.value, next);
		this.transferLinks(q, t);
		return t;
	}


	@Override
	void afterNodeAccess(Node e) {
		synchronized (this.LOCK) {
			Entry last;
			if (this.accessOrder && (last = this.tail) != e) {
				Entry p =
					(Entry) e, b = p.before, a = p.after;
				p.after = null;
				if (b == null) {
					this.head = a;
				}
				else {
					b.after = a;
				}
				if (a != null) {
					a.before = b;
				}
				else {
					last = b;
				}
				if (last == null) {
					this.head = p;
				}
				else {
					p.before = last;
					last.after = p;
				}
				this.tail = p;
			}
		}
	}

	@Override
	public boolean containsValue(Context context, Value<?> value) throws CodeError {
		if (value == null) {
			return false;
		}
		synchronized (this.LOCK) {
			for (Entry e = this.head; e != null; e = e.after) {
				Value<?> v = e.value;
				if (v == value || value.isEquals(context, v)) {
					return true;
				}
			}
			return false;
		}
	}

	@Override
	public Value<?> get(Context context, Value<?> key) throws CodeError {
		synchronized (this.LOCK) {
			Node e = this.getNode(context, key);
			if (e == null) {
				return null;
			}
			if (this.accessOrder) {
				this.afterNodeAccess(e);
			}
			return e.value;
		}
	}

	@Override
	public void clear() {
		synchronized (this.LOCK) {
			super.clear();
			this.head = this.tail = null;
		}
	}

	@SuppressWarnings("unused")
	protected boolean removeEldestEntry(Entry eldest) {
		return false;
	}

	@Override
	public ArucasList keys() {
		ArucasList keyList = new ArucasList();
		synchronized (this.LOCK) {
			for (Entry e = this.head; e != null; e = e.after) {
				keyList.add(e.key);
			}
			return keyList;
		}
	}

	@Override
	public ArucasList values() {
		ArucasList valueList = new ArucasList();
		synchronized (this.LOCK) {
			for (Entry e = this.head; e != null; e = e.after) {
				valueList.add(e.value);
			}
			return valueList;
		}
	}

	@Override
	public Set<ValuePair> pairSet() {
		Set<ValuePair> pairSet = new LinkedHashSet<>();
		synchronized (this.LOCK) {
			for (Entry e = this.head; e != null; e = e.after) {
				pairSet.add(new ValuePair(e.key, e.value));
			}
			return pairSet;
		}
	}

	@Override
	void afterNodeInsertion(Context context, boolean evict) throws CodeError {
		Entry first;
		if (evict && (first = this.head) != null && this.removeEldestEntry(first)) {
			Value<?> key = first.key;
			this.removeNode(context, hash(context, key), key, null, false, true);
		}
	}

	@Override
	void afterNodeRemoval(Node e) {
		Entry p =
			(Entry) e, b = p.before, a = p.after;
		p.before = p.after = null;
		if (b == null) {
			this.head = a;
		}
		else {
			b.after = a;
		}
		if (a == null) {
			this.tail = b;
		}
		else {
			a.before = b;
		}
	}

	static class Entry extends Node {
		Entry before, after;

		Entry(int hash, Value<?> key, Value<?> value, Node next) {
			super(hash, key, value, next);
		}
	}
}

File: src/main/kotlin/me/senseiwells/arucas/utils/impl/ArucasOrderedMap.kt
package me.senseiwells.arucas.utils.impl

import me.senseiwells.arucas.classes.instance.ClassInstance
import me.senseiwells.arucas.interpreter.Interpreter

class ArucasOrderedMap(): ArucasMap() {
    private val accessOrder: Boolean

    private var head: Entry? = null
    private var tail: Entry? = null

    init {
        this.accessOrder = false
    }

    constructor(interpreter: Interpreter, map: ArucasMap): this() {
        this.putMapEntries(interpreter, map, false)
    }

    constructor(interpreter: Interpreter, map: Map<*, *>): this() {
        for ((key, value) in map) {
            this.put(interpreter, interpreter.convertValue(key), interpreter.convertValue(value))
        }
    }

    override fun newNode(hash: Int, key: ClassInstance, value: ClassInstance, next: TreeNode?): Node {
        val entry = Entry(hash, key, value, next)
        linkNodeLast(entry)
        return entry
    }

    override fun replacementNode(node: Node, next: Node?): Node {
        val q = node as Entry
        val t = Entry(q.hash, q.key, q.value, next)
        transferLinks(q, t)
        return t
    }

    override fun newTreeNode(hash: Int, key: ClassInstance, value: ClassInstance, next: Node?): TreeNode {
        val p = TreeNode(hash, key, value, next)
        linkNodeLast(p)
        return p
    }

    override fun replacementTreeNode(node: Node, next: Node?): TreeNode {
        val q = node as Entry
        val t = TreeNode(q.hash, q.key, q.value, next)
        transferLinks(q, t)
        return t
    }

    override fun afterNodeAccess(e: Node) {
        synchronized(lock) {
            var last = tail
            if (accessOrder && last !== e) {
                val p = e as Entry
                val b = p.before
                val a = p.after
                p.after = null
                if (b == null) {
                    head = a
                } else {
                    b.after = a
                }
                if (a != null) {
                    a.before = b
                } else {
                    last = b
                }
                if (last == null) {
                    head = p
                } else {
                    p.before = last
                    last.after = p
                }
                tail = p
            }
        }
    }

    override fun containsValue(interpreter: Interpreter, value: ClassInstance): Boolean {
        synchronized(lock) {
            var e = head
            while (e != null) {
                val v = e.value
                if (v === value || value.equals(interpreter, v)) {
                    return true
                }
                e = e.after
            }
            return false
        }
    }

    override fun get(interpreter: Interpreter, key: ClassInstance): ClassInstance? {
        synchronized(lock) {
            val e = getNode(interpreter, key) ?: return null
            if (accessOrder) {
                afterNodeAccess(e)
            }
            return e.value
        }
    }

    override fun clear() {
        synchronized(lock) {
            super.clear()
            tail = null
            head = null
        }
    }

    override fun keys(): ArucasList {
        val keyList = ArucasList()
        synchronized(lock) {
            var e = head
            while (e != null) {
                keyList.add(e.key)
                e = e.after
            }
            return keyList
        }
    }

    override fun values(): ArucasList {
        val valueList = ArucasList()
        synchronized(lock) {
            var e = head
            while (e != null) {
                valueList.add(e.value)
                e = e.after
            }
            return valueList
        }
    }

    override fun pairSet(): Set<Pair<ClassInstance, ClassInstance>> {
        val pairSet = LinkedHashSet<Pair<ClassInstance, ClassInstance>>()
        synchronized(lock) {
            var e = head
            while (e != null) {
                pairSet.add(e.key to e.value)
                e = e.after
            }
            return pairSet
        }
    }

    override fun afterNodeInsertion(interpreter: Interpreter, evict: Boolean) {

    }

    override fun afterNodeRemoval(p: Node) {
        synchronized(this.lock) {
            val e = p as Entry
            val b = e.before
            val a = e.after
            e.after = null
            e.before = null
            if (b == null) {
                head = a
            } else {
                b.after = a
            }
            if (a == null) {
                tail = b
            } else {
                a.before = b
            }
        }
    }

    private fun linkNodeLast(entry: Entry) {
        synchronized(lock) {
            val last = tail
            tail = entry
            if (last == null) {
                head = entry
                return
            }
            entry.before = last
            last.after = entry
        }
    }

    private fun transferLinks(source: Entry, destination: Entry) {
        synchronized(lock) {
            destination.before = source.before
            val b = destination.before
            destination.after = source.after
            val a = destination.after
            if (b == null) {
                head = destination
            } else {
                b.after = destination
            }
            if (a == null) {
                tail = destination
            } else {
                a.before = destination
            }
        }
    }

    open class Entry(
        hash: Int,
        key: ClassInstance,
        value: ClassInstance,
        next: Node?
    ): Node(hash, key, value, next) {
        var before: Entry? = null
        var after: Entry? = null
    }
}